fastestimator/backend/update_model.py

```diff
@@ -16,117 +16,93 @@
     loss = fe.backend.sparse_categorical_crossentropy(y_pred=pred, y_true=y)  # ~2.3
-    fe.backend.update_model(m, loss=loss)
+    gradients = fe.backend.get_gradient(target=loss,
+                                        sources=[x for x in m.parameters() if x.requires_grad])
+
+    fe.backend.update_model(m, gradients=gradients)
     ```
 
     Args:
         model: A neural network instance to update.
-        loss: A loss value to compute gradients from, mutually exclusive with `gradients`.
-        gradients: A list of tensors to update the models, mutually exclusive with `loss`.
+        gradients: A list of tensors to update the models.
         tape: A TensorFlow GradientTape which was recording when the `loss` was computed (iff using TensorFlow).
         retain_graph: Whether to keep the model graph in memory (applicable only for PyTorch).
-        scaler: A PyTorch loss scaler that scales loss when PyTorch mixed precision is used.
         defer: If True, then the model update function will be stored into the `deferred` dictionary rather than
             applied immediately.
         deferred: A dictionary in which model update functions are stored.
 
     Raises:
         ValueError: If `model` is an unacceptable data type.
-        RuntimeError: If attempting to modify a PyTorch model which relied on gradients within a different PyTorch model
-            which has in turn already undergone a non-deferred update.
+        AssertionError: If `model` doesn't have `current_optimizer` attribute
+        AssertionError: If Pytorch `model.current_optimizer` doesn't have `scaler` attribute
     """
-    if loss is not None:
-        loss = reduce_mean(loss)
+    assert hasattr(model, "current_optimizer"), ("The model needs to have 'current_optimizer' attribute. Please "
+                                                 "instantiate the model with fe.build")
+
     if isinstance(model, tf.keras.Model):
-        if loss is not None:
-            # scale up loss for mixed precision training to avoid underflow
-            if isinstance(model.current_optimizer, mixed_precision.LossScaleOptimizer):
-                loss = model.current_optimizer.get_scaled_loss(loss)
-            # for multi-gpu training, the gradient will be combined by sum, normalize the loss
-            strategy = tf.distribute.get_strategy()
-            if isinstance(strategy, tf.distribute.MirroredStrategy):
-                loss = loss / strategy.num_replicas_in_sync
-            gradients = get_gradient(loss, model.trainable_variables, tape=tape)
-        with tape.stop_recording():
-            # scale down gradient to balance scale-up loss
-            if isinstance(model.current_optimizer, mixed_precision.LossScaleOptimizer):
-                gradients = model.current_optimizer.get_unscaled_gradients(gradients)
+        variables = model.trainable_variables
+        with tape.stop_recording() if tape else NonContext():
```

Comment:
do we still need to stop_recording in order to apply gradients that were already computed?

Reply:
not sure. I will remove it and try.

Reply:
fixed